lib/ansible/modules/system/firewalld_ipset.py

```diff
@@ -0,0 +1,267 @@
+    if indata['permanent']:
+        client.runtimeToPermanent()
+
+
+def run():
+    module_args = dict(
+        name=dict(type='str', required=True),
+        state=dict(choices=['present', 'absent'], required=True),
+        settype=dict(choices=[
+            "hash:ip", "hash:ip,port", "hash:ip,port,ip",
+            "hash:ip,port,net", "hash:ip,mark",
+            "hash:net", "hash:net,net",
+            "hash:net,port", "hash:net,port,net", "hash:net,iface",
+            "hash:mac"
+        ], required=True),
+        permanent=dict(type='bool', required=False, default=False),
+        immediate=dict(type='bool', required=False, default=False),
+        addresses=dict(type='list', required=False)
+    )
+
+    module = AnsibleModule(
+        argument_spec=module_args,
+        supports_check_mode=True
+    )
+
+    # setup the FirewallDClient
+    client = FirewallClient()
+    sets = client.getIPSets()
+    settings = FirewallClientIPSetSettings()
+    settings.setType(module.params['settype'])
+    config = client.config()
+
+    # construct return data
+    result = {
+        "changed": False,
+        "firewalld_ipset_name": module.params['name'],
+        "firewalld_ipset_addresses": module.params['addresses']
+    }
+
+    if module.check_mode:
+        return result
+
+    # Modifying a preexisting ipset
+    if module.params['name'] in sets and module.params['state'] == 'present':
+        client_ipset_config = config.getIPSetByName(module.params['name'])
+        original_entries = client_ipset_config.getEntries()
+        client_ipset_config.setEntries(module.params['addresses'])
+        firewalld_state(client, module.params)
+        new_entries = client_ipset_config.getEntries()
+        result['changed'] = (new_entries != original_entries)
+        result['firewalld_ipset_addresses'] = new_entries
+
+    # Creating a new ipset because the one proposed in the module declaration
+    # does not exist already.
+    elif module.params['name'] not in sets and module.params['state'] == 'present':
+        client_ipset_config = config.addIPSet(module.params['name'], settings)
+        original_entries = client_ipset_config.getEntries()
+        client_ipset_config.setEntries(module.params['addresses'])
+        firewalld_state(client, module.params)
+        new_entries = client_ipset_config.getEntries()
+        result['changed'] = (new_entries != original_entries)
+        result['firewalld_ipset_addresses'] = new_entries
+
+    # Removing an ipset that exists right now. If an ipset is asked to be
+    # removed when it doesn't exist, we should just return an "unchanged"
+    # state.
+    elif module.params['name'] in sets and module.params['state'] == 'absent':
+        client_ipset_config = config.getIPSetByName(module.params['name'])
+        original_entries = client_ipset_config.remove()
+        firewalld_state(client, module.params)
+        result['changed'] = True
+
+    module.exit_json(**result)
+
+
+def main():
```

Comment:
This seems redundant since you only call `run()` here and the in `if __name__ == '__main__`:` you only call `main()`. I'd suggest renaming `run` to `main`.